### ISSUE TITLE
support for sorts & proper generalized arities

### ISSUE THREAD

**Issue by jonsterling** [jonsterling/hs-abt#4]
this one is tricky, but probably necessary to achieve the full generality of ABTs as defined in PFPL.

**Comment by jonsterling** [comment]
(I'll just note, with this addition, the library would be getting closer and closer to being a logical framework in its own right; i.e. abstraction is the primitive function space, opcodes are the signature.)

**Comment by jonsterling** [comment]
I implemented something along these lines yesterday, but I have come to the conclusion that it is probably not worth it for the following reasons:

1. The types get very, very gnarly and it is not so much fun to work with.
2. If I intend this to be used in essentially LCF-style theorem provers where proof scripts are written directly in Haskell, then I suppose having sorts enforced in the Haskell type system makes sense. But my primary use-case is to write typechecker for actual languages and theories, and as such, having sorts just pushes some of the work of type checking into parsing, and I am not sure that that is worth it.
3. Having higher-order variables is pretty neat, but as soon as I do this, it means that the very simple-minded kind of canonicity I have been able to have when working with ABTs (i.e. everything of higher order is an abstraction, etc.) will no longer be possible without having a routine to *normalize ABTs*. This adds a lot of complication, and whilst it is necessary for a proper logical framework, I do not want to deal with it in this library.

For all these reasons, I'm closing this issue unless there is some simplifying notion that arises which convinces me otherwise.
